File: packages/provide-uterm-go/detection/buffer.go
//
// SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
// SPDX-License-Identifier: AGPL-3.0-or-later
//

package detection

import "time"

// nowSeconds returns the current wall-clock time as fractional Unix seconds,
// matching Python's time.time(). It is a package var so tests can pin it.
var nowSeconds = func() float64 {
	return float64(time.Now().UnixNano()) / 1e9
}

// ScreenBuffer is a buffered screen snapshot with timing metadata. Faithful
// to the Python pydantic model.
type ScreenBuffer struct {
	Screen              string
	ScreenHash          string
	Snapshot            Snapshot
	CapturedAt          float64
	MatchedPromptID     string
	TimeSinceLastChange float64
}

// BufferManager manages a screen-history buffer with timing calculation.
type BufferManager struct {
	maxSize        int
	buffer         []*ScreenBuffer
	lastHash       string
	lastChangeTime float64
}

// NewBufferManager creates a buffer manager retaining at most maxSize screens.
func NewBufferManager(maxSize int) *BufferManager {
	return &BufferManager{maxSize: maxSize}
}

// AddScreen adds a screen snapshot to the buffer and calculates its timing
// metadata, returning the created ScreenBuffer.
func (m *BufferManager) AddScreen(snapshot Snapshot) *ScreenBuffer {
	now := nowSeconds()
	if v, ok := snapshot["captured_at"]; ok {
		if f, fok := toFloat(v); fok {
			now = f
		}
	}
	screenHash, _ := snapshot["screen_hash"].(string)

	var timeSinceChange float64
	if m.lastChangeTime > 0 {
		timeSinceChange = now - m.lastChangeTime
	}
	if screenHash != m.lastHash {
		m.lastHash = screenHash
		m.lastChangeTime = now
	}

	screenText, _ := snapshot["screen"].(string)
	buf := &ScreenBuffer{
		Screen:              screenText,
		ScreenHash:          screenHash,
		Snapshot:            snapshot,
		CapturedAt:          now,
		TimeSinceLastChange: timeSinceChange,
	}
	m.buffer = append(m.buffer, buf)
	if len(m.buffer) > m.maxSize {
		m.buffer = m.buffer[len(m.buffer)-m.maxSize:]
	}
	return buf
}

// GetRecent returns the n most recent buffered screens (oldest first). When
// n is >= the buffer length, all buffered screens are returned.
func (m *BufferManager) GetRecent(n int) []*ScreenBuffer {
	if n >= len(m.buffer) {
		out := make([]*ScreenBuffer, len(m.buffer))
		copy(out, m.buffer)
		return out
	}
	out := make([]*ScreenBuffer, n)
	copy(out, m.buffer[len(m.buffer)-n:])
	return out
}

// DetectIdleState reports whether the screen has been unchanged for at least
// thresholdSeconds. A fresh manager (no screens) is never idle.
func (m *BufferManager) DetectIdleState(thresholdSeconds float64) bool {
	if m.lastChangeTime == 0 || m.lastHash == "" {
		return false
	}
	return (nowSeconds() - m.lastChangeTime) >= thresholdSeconds
}

// Len returns the number of buffered screens.
func (m *BufferManager) Len() int { return len(m.buffer) }

// MaxSize returns the configured maximum buffer size.
func (m *BufferManager) MaxSize() int { return m.maxSize }

// Clear empties the buffer and resets change tracking.
func (m *BufferManager) Clear() {
	m.buffer = nil
	m.lastHash = ""
	m.lastChangeTime = 0
}

// toFloat coerces a decoded-JSON numeric value to float64.
func toFloat(v any) (float64, bool) {
	switch x := v.(type) {
	case float64:
		return x, true
	case int:
		return float64(x), true
	case int64:
		return float64(x), true
	default:
		return 0, false
	}
}

File: packages/provide-uterm-ts/src/detection/buffer.ts
//
// SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
// SPDX-License-Identifier: AGPL-3.0-or-later
//

/**
 * Screen buffering with timing metadata for prompt detection.
 *
 * Port of the Python module `provide.uterm.detection.buffer` and the Go
 * package `detection`.
 */

/**
 * A screen snapshot, as the emulator hands one over.
 *
 * The reference declares this once, in `detection.models`, and both the
 * buffer and the detector read it: the screen and its hash are required, and
 * the cursor metadata is whatever the producer could work out. The detector
 * takes a `Partial` of it, because it provably survives a snapshot that is
 * missing the required fields and turning that into a crash mid-frame would
 * end a session over a malformed frame.
 */
export interface ScreenSnapshot {
  screen: string;
  screen_hash: string;
  /** Capture time in seconds. Defaults to now when absent. */
  captured_at?: number;
  /** Whether the cursor sits after the last drawn character. */
  cursor_at_end?: boolean;
  /** Whether the screen ends in a space, which suggests a live input field. */
  has_trailing_space?: boolean;
  /** Where the cursor is, as far as the emulator knows. */
  cursor?: { x?: unknown; y?: unknown } | null;
  [key: string]: unknown;
}

/** A buffered screen snapshot with timing metadata. */
export interface ScreenBuffer {
  screen: string;
  screen_hash: string;
  snapshot: ScreenSnapshot;
  captured_at: number;
  matched_prompt_id: string | null;
  /** Seconds since the screen last changed, or 0 before the first change. */
  time_since_last_change: number;
}

/** Manages a bounded screen history with timing calculation. */
export class BufferManager {
  readonly #maxSize: number;
  #buffer: ScreenBuffer[] = [];
  #lastHash = "";
  #lastChangeTime = 0;

  constructor(maxSize = 50) {
    this.#maxSize = maxSize;
  }

  /** How many screens are held. */
  get size(): number {
    return this.#buffer.length;
  }

  /** The most this will hold before dropping the oldest. */
  get maxSize(): number {
    return this.#maxSize;
  }

  /**
   * Add a screen snapshot and compute its timing metadata.
   *
   * The elapsed time is measured from the last *change*, not the last
   * snapshot, so a screen that keeps arriving unchanged reports a growing
   * idle interval. It stays zero until a first change has been seen, which
   * is what keeps a freshly-started session from looking idle.
   */
  addScreen(snapshot: ScreenSnapshot): ScreenBuffer {
    const now = snapshot.captured_at ?? Date.now() / 1000;
    const screenHash = snapshot.screen_hash;
    const elapsed = this.#lastChangeTime > 0 ? now - this.#lastChangeTime : 0;

    if (screenHash !== this.#lastHash) {
      this.#lastHash = screenHash;
      this.#lastChangeTime = now;
    }

    const buffer: ScreenBuffer = {
      screen: snapshot.screen,
      screen_hash: screenHash,
      snapshot,
      captured_at: now,
      matched_prompt_id: null,
      time_since_last_change: elapsed,
    };
    this.#buffer.push(buffer);
    if (this.#buffer.length > this.#maxSize) {
      this.#buffer = this.#buffer.slice(-this.#maxSize);
    }
    return buffer;
  }

  /** The `n` most recent buffered screens, oldest first. */
  getRecent(n = 5): ScreenBuffer[] {
    if (n >= this.#buffer.length) {
      return [...this.#buffer];
    }
    return this.#buffer.slice(-n);
  }

  /**
   * Whether the screen has been unchanged for at least `thresholdSeconds`.
   *
   * Measured against the wall clock rather than the last snapshot, so a
   * session that has stopped producing output goes idle without needing
   * another snapshot to arrive and say so.
   */
  detectIdleState(thresholdSeconds = 2): boolean {
    if (this.#lastChangeTime === 0 || this.#lastHash === "") {
      return false;
    }
    return Date.now() / 1000 - this.#lastChangeTime >= thresholdSeconds;
  }

  /** Clear the buffer and reset the change clock. */
  clear(): void {
    this.#buffer = [];
    this.#lastHash = "";
    this.#lastChangeTime = 0;
  }
}
